eureka2-client/src/main/java/com/netflix/eureka2/client/interest/InterestHandlerImpl.java

```diff
@@ -17,33 +19,106 @@
+
+    private final RetryableConnection<InterestChannel> retryableConnection;
 
     @Inject
     public InterestHandlerImpl(SourcedEurekaRegistry<InstanceInfo> registry,
-                               ClientChannelFactory<ClientInterestChannel> channelFactory) {
+                               ChannelFactory<InterestChannel> channelFactory) {
+        this(registry, channelFactory, DEFAULT_RETRY_WAIT_MILLIS);
+    }
+
+    /* visible for testing*/ InterestHandlerImpl(SourcedEurekaRegistry<InstanceInfo> registry,
+                                                 ChannelFactory<InterestChannel> channelFactory,
+                                                 int retryWaitMillis) {
         this.registry = registry;
-        this.interestChannel = channelFactory.newChannel();
+        this.interestTracker = new InterestTracker();
         this.isShutdown = new AtomicBoolean(false);
+
+        RetryableConnectionFactory<InterestChannel, Interest<InstanceInfo>> retryableConnectionFactory
+                = new RetryableConnectionFactory<InterestChannel, Interest<InstanceInfo>>(channelFactory) {
+                    @Override
+                    protected Observable<Void> executeOnChannel(InterestChannel channel, Interest<InstanceInfo> interest) {
+                        return channel.change(interest);
+                    }
+                };
+
+        this.retryableConnection = retryableConnectionFactory.newConnection(interestTracker.interestChangeStream());
+        retryableConnection.getRetryableLifecycle()
+                .retryWhen(new RetryStrategyFunc(retryWaitMillis))
+                .subscribe(new Subscriber<Void>() {
+                    @Override
+                    public void onCompleted() {
+                        logger.info("channel onCompleted");
+                    }
+
+                    @Override
+                    public void onError(Throwable e) {
+                    }
+
+                    @Override
+                    public void onNext(Void aVoid) {
+
+                    }
+                });
+    }
+
+    /**
+     * TODO: make a final decision on:
+     * if channelLifecycle retries enough times and does not succeed, do we indefinitely retry and/or propagate
+     * the error to all the subscribers of forInterests? We could also do this through the registry index.
+     */
+    @Override
+    public Observable<ChangeNotification<InstanceInfo>> forInterest(final Interest<InstanceInfo> interest) {
+        if (isShutdown.get()) {
+            return Observable.error(new IllegalStateException("InterestHandler has shutdown"));
+        }
+
+        Observable<Void> appendInterest = Observable.create(new Observable.OnSubscribe<Void>() {
+            @Override
+            public void call(Subscriber<? super Void> subscriber) {
+                interestTracker.appendInterest(interest);
+                subscriber.onCompleted();
+            }
+        });
+
+        Observable toReturn = appendInterest
```

Comment:
Should we merge this with lifecycle, so in case retries fail, ultimately we notify the client about an error. The only other way to do that would be to shutdown registry with an error, but we do not want to do that.